br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,254 @@
+			lastOffset, _ = fileParser.Pos()
+		}
+
+		fileParser.Close()
+
+		d.totalRows.Add(rowsChecked)
+		d.totalColumnCountMismatchRows.Add(columnCountMismatchRows)
+		d.totalInvalidCharRows.Add(invalidCharRows)
+	}
+}
+
+func (d *dataSampleCheck) doCheck(ctx context.Context) error {
+	rc := d.controller
+
+	targetFiles, err := d.getSampledDataFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	concurrency := utils.MinInt(runtime.NumCPU(), 8)
+	fileChan := make(chan *sampledDataFileInfo, concurrency)
+	errChan := make(chan error, concurrency)
+	newCtx, cancelFunc := context.WithCancel(ctx)
+
+	for i := 0; i < concurrency; i++ {
+		d.wg.Add(1)
+		go d.checkRoutine(newCtx, fileChan, errChan)
+	}
+
+	for _, fileInfo := range targetFiles {
+		fileChan <- fileInfo
+	}
+	close(fileChan)
+
+	// we need to exit if any error occurs in check routine
+	// each routine should return a result in the error channel, return nil if there's no error.
+	for i := 0; i < concurrency; i++ {
+		if err = <-errChan; err != nil {
+			cancelFunc()
+			d.wg.Wait()
+
+			log.L().Error("meet error when checking", zap.Error(err))
+			fmt.Printf("failed to do data sample check: %s\n", err.Error())
+			return errors.Trace(err)
+		}
+	}
+
+	d.wg.Wait()
+	cancelFunc()
+
+	passed := d.totalColumnCountMismatchRows.Load() == 0 && d.totalInvalidCharRows.Load() == 0
+	msg := fmt.Sprintf("Total sample of %d rows of data checked, %d errors found.",
+		d.totalRows.Load(), d.totalColumnCountMismatchRows.Load()+d.totalInvalidCharRows.Load())
+	d.checkTemplate.Collect(Critical, passed, msg)
+
+	if rc.tidbGlue.OwnsSQLExecutor() {
+		fmt.Println(d.checkTemplate.Output())
+
+		if d.checkTemplate.Success() {
+			fmt.Println("All checks have been passed, but there may still be other types of errors that can only be found during the actual insertion of data.")
+		} else {
+			fmt.Println("Some checks failed, please check the log for more information.")
+			fmt.Printf("Log file location: %s\n", rc.cfg.LogCfg.File)
+		}
+	}
+
+	return nil
+}
+
+type sampledDataFileInfo struct {
+	DataFile  *mydump.FileInfo
+	Table     *mydump.MDTableMeta
+	TableInfo *model.TableInfo
+}
+
+func (d *dataSampleCheck) getSampledDataFiles() ([]*sampledDataFileInfo, error) {
+	rc := d.controller
+
+	dataFiles := d.getRandomDataFiles()
+	for _, fileInfo := range dataFiles {
+		tableMeta := fileInfo.Table
+
+		tableInfo := rc.dbInfos[tableMeta.DB].Tables[tableMeta.Name]
+		fileInfo.TableInfo = tableInfo.Core
+	}
+	return dataFiles, nil
+}
+
+func (d *dataSampleCheck) getRandomDataFiles() []*sampledDataFileInfo {
+	allFiles := make([]*sampledDataFileInfo, 0)
+
+	rc := d.controller
+	for _, dbInfo := range rc.dbMetas {
+		for _, tableInfo := range dbInfo.Tables {
+			for i := range tableInfo.DataFiles {
+				allFiles = append(allFiles, &sampledDataFileInfo{
+					DataFile: &tableInfo.DataFiles[i],
+					Table:    tableInfo,
+				})
+			}
+		}
+	}
+
+	count := int(float64(len(allFiles)) * d.checkCfg.Rate)
```

Comment:
I think check per file level is not a quite good idea, in case the total file number is not very big but each file is relative big in size, this random logic will result in quite slow speed. How about check the first N bytes of each file?

Reply:
> How about check the first N bytes of each file?

when there's a large number of files, it's better check a random set of them.

Reply:
Besides, if there are many tables and one of the table has more files than others, this check may skip most of the tables. However, sampling more tables seems more better than sampling more files.  I suggest sampling every table and perform random file choosing every table.

Reply:
@sunzhaoyang have a look